File: shared/idempotency.js
const redisModule = require('./redis');

const TTL_SEC = 86400; // 24h — aligns with API doc

async function alreadyProcessed(idempotencyKey) {
  if (!idempotencyKey) return false;
  await redisModule.connectRedis();
  const k = `idem:${idempotencyKey}`;
  const v = await redisModule.client.get(k);
  return !!v;
}

async function markProcessed(idempotencyKey) {
  if (!idempotencyKey) return;
  await redisModule.connectRedis();
  await redisModule.client.setEx(`idem:${idempotencyKey}`, TTL_SEC, '1');
}

module.exports = { alreadyProcessed, markProcessed };

File: backend/app/idempotency.py
from __future__ import annotations

TTL_SEC = 86400


async def already_processed(idempotency_key: str | None, prefix: str = "") -> bool:
    if not idempotency_key:
        return False
    try:
        from app.redis_client import get_redis

        r = get_redis()
        k = f"idem:{prefix}{idempotency_key}"
        v = await r.get(k)
        return bool(v)
    except Exception:
        return False


async def mark_processed(idempotency_key: str | None, prefix: str = "") -> None:
    if not idempotency_key:
        return
    try:
        from app.redis_client import get_redis

        r = get_redis()
        await r.setex(f"idem:{prefix}{idempotency_key}", TTL_SEC, "1")
    except Exception:
        pass
